File: backend/src/Axle.WebApi/Models/Tenant.cs
using System.ComponentModel.DataAnnotations;

namespace Axle.Models;

/// <summary>
/// Represents a tenant in the multi-tenant system.
/// Each tenant has isolated data and custom field configurations.
/// </summary>
public class Tenant
{
    /// <summary>
    /// Unique identifier for the tenant.
    /// </summary>
    public Guid Id { get; set; } = Guid.NewGuid();

    /// <summary>
    /// Display name of the tenant/organization.
    /// </summary>
    [Required]
    [MaxLength(200)]
    public string Name { get; set; } = string.Empty;

    /// <summary>
    /// URL-friendly slug for the tenant (e.g., acme-corp).
    /// </summary>
    [Required]
    [MaxLength(100)]
    public string Slug { get; set; } = string.Empty;

    /// <summary>
    /// Whether the tenant is active and can be accessed.
    /// </summary>
    public bool IsActive { get; set; } = true;

    /// <summary>
    /// JSON blob for tenant-specific settings and configuration.
    /// </summary>
    public string? Settings { get; set; }

    /// <summary>
    /// When the tenant was created.
    /// </summary>
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

    /// <summary>
    /// When the tenant was last modified.
    /// </summary>
    public DateTime? UpdatedAt { get; set; }

    /// <summary>
    /// Users associated with this tenant.
    /// </summary>
    public ICollection<TenantUser> TenantUsers { get; set; } = new List<TenantUser>();

    /// <summary>
    /// Workspaces belonging to this tenant.
    /// </summary>
    public ICollection<Workspace> Workspaces { get; set; } = new List<Workspace>();

    /// <summary>
    /// Nodes belonging to this tenant.
    /// </summary>
    public ICollection<Node> Nodes { get; set; } = new List<Node>();

    /// <summary>
    /// Field definitions for this tenant.
    /// </summary>
    public ICollection<FieldDefinition> FieldDefinitions { get; set; } = new List<FieldDefinition>();
}

File: frontend/flutter/lib/domain/models/tenant.dart
/// Represents a tenant in the multi-tenant system.
class Tenant {
  /// Unique identifier for the tenant.
  final String id;

  /// Display name of the tenant/organization.
  final String name;

  /// URL-friendly slug for the tenant (e.g., acme-corp).
  final String slug;

  /// Whether the tenant is active and can be accessed.
  final bool isActive;

  /// JSON blob for tenant-specific settings and configuration.
  final String? settings;

  /// When the tenant was created.
  final DateTime createdAt;

  /// When the tenant was last modified.
  final DateTime? updatedAt;

  const Tenant({
    required this.id,
    required this.name,
    required this.slug,
    this.isActive = true,
    this.settings,
    required this.createdAt,
    this.updatedAt,
  });

  /// Creates a Tenant from JSON data.
  factory Tenant.fromJson(Map<String, dynamic> json) {
    return Tenant(
      id: json['id'] as String,
      name: json['name'] as String,
      slug: json['slug'] as String,
      isActive: json['isActive'] as bool? ?? true,
      settings: json['settings'] as String?,
      createdAt: DateTime.parse(json['createdAt'] as String),
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt'] as String)
          : null,
    );
  }

  /// Converts the Tenant to JSON.
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'slug': slug,
      'isActive': isActive,
      'settings': settings,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt?.toIso8601String(),
    };
  }

  /// Creates a copy of this Tenant with updated fields.
  Tenant copyWith({
    String? id,
    String? name,
    String? slug,
    bool? isActive,
    String? settings,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Tenant(
      id: id ?? this.id,
      name: name ?? this.name,
      slug: slug ?? this.slug,
      isActive: isActive ?? this.isActive,
      settings: settings ?? this.settings,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}

/// DTO for creating a new tenant.
class CreateTenantDto {
  final String name;
  final String slug;

  const CreateTenantDto({
    required this.name,
    required this.slug,
  });

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'slug': slug,
    };
  }
}
